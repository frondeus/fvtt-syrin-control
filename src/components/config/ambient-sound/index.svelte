<script lang="ts">
	import Context from '@/services/context';
	import type { SyrinAmbientSoundFlags } from '@/sounds';
	import type { Soundsets } from '@/models/store';

	// Context
	const ctx = Context();

	// Params & State
	interface Darkness {
		min: number;
		max: number;
	}

	export let x: number;
	export let y: number;
	export let radius: number;
	export let walls: boolean;
	export let darkness: Darkness;

	export let flags: SyrinAmbientSoundFlags;
	export let create: boolean;
	let soundsets = ctx.stores.soundsets;
	let soundsetName: string = '...';
	let ambientName: string = '...';
	let style = '';
	let class_ = 'inner';

	const reactiveAmbientSoundName = async (flags: SyrinAmbientSoundFlags, soundsets: Soundsets) => {
		if (flags.type === 'mood') {
			const soundset = await ctx.stores.hydrateSoundsetInner(
				(await ctx.api.soundsetIdForMood(Number(flags.mood))) ?? '',
				soundsets
			);

			const mood = soundset.moods[flags.mood];
			ambientName = mood.name;
			soundsetName = soundset.name;
			if (soundset.artworkUrl !== undefined) {
				style = `background-image: url('${soundset.artworkUrl}');`;
				class_ = 'inner inner-invert';
			}
		}
	};

	$: reactiveAmbientSoundName(flags, $soundsets);
</script>

<div class={class_} {style}>
	<div>
		<p class="notes">
			{ctx.game.localizeCore('SOUND.ConfigHint')}
		</p>

		<div class="form-group">
			<label for="">{ctx.game.localize('config.soundset')}</label>
			<input type="text" disabled value={soundsetName} title={soundsetName} />
		</div>
		<div class="form-group">
			<label for="">
				{ctx.game.localize('config.' + flags.type)}
			</label>
			<div class="form-fields">
				<input type="text" disabled value={ambientName} title={ambientName} />
			</div>
		</div>
		<div class="form-group">
			<label for=""
				>{ctx.game.localizeCore('XCoord')}
				<span class="units">({ctx.game.localizeCore('Pixels')})</span></label
			>
			<div class="form-fields">
				<input type="number" name="x" step="1" bind:value={x} />
			</div>
		</div>
		<div class="form-group">
			<label for=""
				>{ctx.game.localizeCore('YCoord')}
				<span class="units">({ctx.game.localizeCore('Pixels')})</span></label
			>
			<div class="form-fields">
				<input type="number" name="y" step="1" bind:value={y} />
			</div>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('SOUND.Radius')}</label>
			<div class="form-fields">
				<input type="number" name="radius" step="any" bind:value={radius} />
			</div>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('SOUND.Walls')}</label>
			<input type="checkbox" name="walls" bind:checked={walls} />
			<p class="hint">
				{ctx.game.localizeCore('SOUND.WallsHint')}
			</p>
		</div>
		<div class="form-group">
			<label for="">{ctx.game.localizeCore('SOUND.DarknessRange')}</label>
			<div class="form-fields">
				<label for="darkness.min">{ctx.game.localizeCore('Between')}</label>
				<input
					type="number"
					name="darkness.min"
					value={darkness.min}
					min="0"
					max="1"
					step="any"
					placeholder="0"
				/>
				<label for="darkness.max">{ctx.game.localizeCore('and')}</label>
				<input
					type="number"
					name="darkness.max"
					value={darkness.max}
					min="0"
					max="1"
					step="any"
					placeholder="1"
				/>
			</div>
			<p class="hint">
				{ctx.game.localizeCore('SOUND.DarknessRangeHint')}
			</p>
		</div>
	</div>
	<input type="hidden" name="path" value="syrinscape.wav" />
	<input type="hidden" name="flags.syrinscape.type" value={flags.type} />
	{#if flags.type === 'mood'}
		<input type="hidden" name="flags.syrinscape.mood" value={flags.mood} />
	{/if}
	<div>
		{ctx.game.localize('config.controlled', { name: ctx.game.localize('config.ambientSound') })}
		<button type="submit">
			<i class="far fa-save" />
			{#if create}
				{ctx.game.localizeCore('SOUND.Create')}
			{:else}
				{ctx.game.localizeCore('SOUND.Update')}
			{/if}
		</button>
	</div>
</div>

<style>
	.inner {
		padding: 8px;
		display: flex;
		flex-direction: column;
		justify-content: space-between;
		height: 100%;
	}
	.inner-invert {
		color: var(--color-text-light-highlight);
		background-size: cover;
	}
	.inner-invert .notes,
	.inner-invert .hint,
	.inner-invert .units {
		color: var(--color-text-light-highlight);
	}
	.inner button,
	.inner input {
		background: url(../ui/parchment.jpg) repeat;
	}
</style>
