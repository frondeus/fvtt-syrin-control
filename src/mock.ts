import { RawApi } from '@/services/raw';
import { FVTTGame } from '@/services/game';
import { storeDefaults, Stores } from '@/services/stores';
import { Utils } from './services/utils';
import { Api } from './services/api';
import { Syrin } from './services/syrin';
import { Context } from './services/context';

export function mockedApi(): RawApi {
	return {
		stopMood: jest.fn(),
		playMood: jest.fn(),
		playElement: jest.fn(),
		getMoods: jest.fn(async (_) => []),
		getElements: jest.fn(async (_) => []),
		getSoundsets: jest.fn(async () => []),
		getGlobalElements: jest.fn(async () => []),
		onInit: jest.fn(),
		getState: jest.fn(),
		changePlayerVolume: jest.fn(),
		changeMoodVolume: jest.fn(),
		changeOneShotVolume: jest.fn(),
		playerJoined: jest.fn(),
		stopElement: jest.fn(),
		getMood: jest.fn()
	};
}

type StoreDefaultsKey = keyof typeof storeDefaults;
export function mockedGame(): FVTTGame {
	return {
		getSetting: jest.fn((name) => {
			if (Object.hasOwn(storeDefaults as object, name)) {
				return storeDefaults[name as StoreDefaultsKey] as any;
			}
			console.error('get setting', { name });
			return null!;
		}),
		setSetting: jest.fn(),
		setGlobal: jest.fn(),
		localize: jest.fn((key) => key),
		registerSetting: jest.fn(),
		notifyInfo: jest.fn(),
		notifyError: jest.fn(),
		callHookAll: jest.fn(),
		getActiveScene: jest.fn(),
		isGM: jest.fn(),
		setSettingToDefault: jest.fn(),
		socket: {
			register: jest.fn(),
			executeAsGM: jest.fn(),
			executeAsUser: jest.fn(),
			executeForAllGMs: jest.fn(),
			executeForEveryone: jest.fn(),
			executeForOtherGMs: jest.fn(),
			executeForOthers: jest.fn(),
			executeForUsers: jest.fn()
		},
		isReady: jest.fn(() => true),
		userId: jest.fn(),
		hasActiveModule: jest.fn(),
		localizeCore: jest.fn(),
		getAudioContext: jest.fn(),
		createMoodMacro: jest.fn(),
		createElementMacro: jest.fn(),
		createPlaylist: jest.fn(),
		createPlaylistSound: jest.fn(),
		createPlaylistMoodSound: jest.fn(),
		getPlaylists: jest.fn(),
		getAmbientSounds: jest.fn(),
		getPlayerName: jest.fn(),
		createDialog: jest.fn()
	};
}

export function mocked(game: FVTTGame = mockedGame(), raw: RawApi = mockedApi()): Context {
	const utils = new Utils(game);
	const api = new Api(raw);
	const stores = new Stores(game, api);
	const syrin = new Syrin(game, utils, api, stores);
	return new Context(game, utils, api, stores, syrin);
}
