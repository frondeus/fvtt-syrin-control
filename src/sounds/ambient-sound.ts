import { Context } from '@/services/context';
import { ElementSoundFlags, MoodSoundFlags } from '@/sounds';

export type SyrinAmbientSoundFlags = ElementSoundFlags | MoodSoundFlags;

export interface AmbientSoundProvider {
	id(): string | null;
	radius(): number;
}
export class AmbientSound {
	ctx: Context;
	flags: SyrinAmbientSoundFlags;
	provider: AmbientSoundProvider;

	constructor(data: any, ctx: Context, provider: AmbientSoundProvider) {
		this.ctx = ctx;
		this.flags = data.flags.syrinscape;
		this.provider = provider;
		this.ctx.utils.trace('Creating an ambient sound', { data, flags: this.flags });
	}

	async sync(isAudible: boolean, volume: number): Promise<void> {
		if (!this.ctx.api.isPlayerActive()) {
			return;
		}

		const id = this.provider.id();
		if (id === null) {
			return;
		}

		const power = (1.0 - volume) * this.provider.radius();
		const userId = this.ctx.game.userId() ?? '';

		if (isAudible) {
			switch (this.flags.type) {
				case 'mood': {
					const moodId = this.flags.mood;
					this.ctx.syrin.playAmbientSound(id, {
						kind: 'mood',
						volume: power,
						moodId,
						userId
					});
					break;
				}
				case 'element': {
					const elementId = this.flags.element;
					this.ctx.syrin.playAmbientSound(id, {
						kind: 'element',
						volume: power,
						elementId,
						userId
					});
					break;
				}
			}
		} else {
			this.ctx.syrin.stopAmbientSound(id, userId);
		}
	}
}
