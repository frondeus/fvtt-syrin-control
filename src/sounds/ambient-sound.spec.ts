import { Mocked, mocked } from '@/mock';
import { AmbientSound, SyrinAmbientSoundFlags } from './ambient-sound';

describe('Ambient sound controller', () => {
	let mock: Mocked;
	let flags: SyrinAmbientSoundFlags;

	beforeEach(() => {
		mock = mocked();
		mock.game.isGM = jest.fn(() => false);
		mock.raw.getState = jest.fn(() => 'active');

		flags = {
			type: 'mood',
			mood: 4321
		};
	});

	it('does nothing when player is not active', async () => {
		const provider = {
			id: () => '1',
			radius: () => 10
		};
		const sound = new AmbientSound({ flags: { syrinscape: flags } }, mock.ctx, provider);

		mock.raw.getState = jest.fn(() => 'inactive');

		await sound.sync(true, 1);

		expect(mock.game.socket?.executeAsGM).not.toBeCalled();
	});

	it('also does nothing when id is unknown', async () => {
		const provider = {
			id: () => null,
			radius: () => 10
		};
		const sound = new AmbientSound({ flags: { syrinscape: flags } }, mock.ctx, provider);

		await sound.sync(true, 1);

		expect(mock.game.socket?.executeAsGM).not.toBeCalled();
	});
});
