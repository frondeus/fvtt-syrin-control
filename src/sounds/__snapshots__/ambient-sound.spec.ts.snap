// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ambient sound controller bugs can be flattened 1`] = `
{
  "flags.mood": 4321,
  "flags.type": "mood",
  "provider.ctx": [Function],
  "provider.id": [Function],
  "provider.radius": [Function],
}
`;
